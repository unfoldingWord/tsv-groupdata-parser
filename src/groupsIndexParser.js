import fs from 'fs-extra';
import path from 'path-extra';
import { getGroupName } from './helpers/resourcesHelpers';

/**
 * make sure this groupId is in a category
 * @param {String} groupId
 * @param {String} groupName - localized group name
 * @param {Object} categorizedGroupsIndex
 * @param {String} categoryName - category that should contain this groupId
 */
const addGroupToCategory = (groupId, groupName, categorizedGroupsIndex, categoryName) => {
  const groupIndexItem = getGroupIndex(groupId, groupName);

  // Only add the groupIndexItem if it isn't already in the category's groups index.
  if (!categorizedGroupsIndex[categoryName].some(e => e.id === groupIndexItem.id)) {
    categorizedGroupsIndex[categoryName].push(groupIndexItem); // adding group Index Item
  }
};

export const generateGroupsIndex = (tnCategoriesPath, taCategoriesPath) => {
  const categorizedGroupsIndex = {
    discourse: [],
    numbers: [],
    figures: [],
    culture: [],
    grammar: [],
    other: [],
  };
  const errors = [];

  const isDirectory = item => fs.lstatSync(path.join(tnCategoriesPath, item)).isDirectory();
  const categories = fs.readdirSync(tnCategoriesPath).filter(isDirectory);

  categories.forEach(categoryName => {
    const booksPath = path.join(tnCategoriesPath, categoryName, 'groups');
    const books = fs.readdirSync(booksPath).filter(item => item !== '.DS_Store');

    books.forEach(bookid => {
      const groupDataPath = path.join(booksPath, bookid);
      const groupDataFiles = fs.readdirSync(groupDataPath).filter(filename => path.extname(filename) === '.json');
      let taArticleCategory;
      let groupName;
      let contextId;

      groupDataFiles.forEach(groupDataFile => {
        try {
          const filePath = path.join(groupDataPath, groupDataFile);
          const groupId = groupDataFile.replace('.json', '');
          const groupData = fs.readJsonSync(filePath);
          taArticleCategory = null;
          groupName = null;
          let foundLocalization = false;
          contextId = null;

          if (groupData.length > 0) {
            for (let i = 0; i < groupData.length; i++ ) {
              const contextId = groupData[i].contextId;
              taArticleCategory = getArticleCategory(contextId.occurrenceNote, groupId);

              try {
                if (!taArticleCategory) {
                  throw new Error(`Link in Occurrence Note '${contextId.occurrenceNote}' does not have category for check at contextID: ${JSON.stringify(contextId)}`);
                }

                const fileName = groupId + '.md';
                const articlePath = path.join(taCategoriesPath, taArticleCategory, fileName);
                groupName = getGroupName(articlePath);
                addGroupToCategory(groupId, groupName, categorizedGroupsIndex, categoryName);
                foundLocalization = true;
                break; // we got the category, so don't need to search anymore
              } catch (e) {
                let message = `error finding group name: groupId: '${groupId}', index: '${i}' in bookId '${bookid}, taArticleCategory: ${taArticleCategory}, contextID: ${JSON.stringify(contextId)}' `;
                console.error('generateGroupsIndex() - ' + message, e);
              }
            }

            if (!foundLocalization) {
              addGroupToCategory(groupId, groupId, categorizedGroupsIndex, categoryName); // add entry even though we could not find localized description
              console.error(`Could not find localization for '${groupId}' in '${categoryName}', adding stub entry`);
            }
          }
        } catch (e) {
          let message = `error processing entry: bookid: ${bookid}, groupDataFile: ${groupDataFile}, taArticleCategory: ${taArticleCategory}, groupName: ${groupName}, contextID: ${JSON.stringify(contextId)}: `;
          console.error('generateGroupsIndex() - ' + message, e);
          errors.push(message + e.toString());
        }
      });
    });
  });

  if (errors.length) {
    throw new Error(`generateGroupsIndex() - error processing index:\n${errors.join('\n')}`);
  }

  return categorizedGroupsIndex;
};

/**
 * Gets the category of the given groupId from the links in the occurrenceNote
 * @param {string} occurrenceNote
 * @param {string} groupId
 * @returns {string} the matched category
 */
export const getArticleCategory = (occurrenceNote, groupId) => {
  if (occurrenceNote && groupId) {
    const pattern = '(?<=rc:\\/\\/[^\\/]+\\/ta\\/man\\/)[^\\/]+?(?=\\/' + groupId + ')';
    const categoryRE = new RegExp(pattern);
    const match = occurrenceNote.match(categoryRE);

    if (match) {
      return match[0];
    }
  }
  return null;
};

const getGroupIndex = (groupId, groupName) => ({
  id: groupId,
  name: groupName,
});
