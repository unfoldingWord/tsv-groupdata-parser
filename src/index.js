import ManageResource from './helpers/ManageResourceAPI';
import * as verseHelpers from './helpers/verseHelpers';
import * as wordOccurrenceHelpers from './helpers/wordOccurrenceHelpers';
export {
  ManageResource,
  verseHelpers,
  wordOccurrenceHelpers,
};
export {
  cleanGroupId,
  generateGroupDataItem,
  tsvToGroupData,
  tnJsonToGroupData,
} from './tsvToGroupData';
export { convertTsv9to7 } from './tsvConversion';
export { generateGroupsIndex } from './groupsIndexParser';
export {
  getCategoryForGroupId, categorizeGroupData, tNotesCategories,
} from './tNoteGroupIdCategorization';
export { formatAndSaveGroupData, saveGroupsIndex } from './helpers/filesystemHelpers';
