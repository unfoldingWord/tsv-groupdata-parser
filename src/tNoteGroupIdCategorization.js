/**
 * Hard coded list of tNotes categories
 */
export const tNotesCategories = {
  discourse: {
    'figs-declarative': 'Statements - Other Uses',
    'figs-events': 'Order of Events',
    'figs-exclamations': 'Exclamations',
    'figs-exmetaphor': 'Extended Metaphor',
    'figs-imperative': 'Imperatives - Other Uses',
    'figs-parables': 'Parables',
    'figs-pastforfuture': 'Predictive Past',
    'figs-quotations': 'Direct and Indirect Quotations',
    'figs-quotesinquotes': 'Quotes Within Quotes',
    'figs-sentences': 'Sentence Structure',
    'translate-versebridge': 'Verse Bridges',
    'writing-background': 'Background Information',
    'writing-connectingwords': 'Connecting Words',
    'writing-endofstory': 'End of Story',
    'writing-intro': 'Types of Writing',
    'writing-newevent': 'Introduction of a New Event',
    'writing-participants': 'Introduction of New and Old Participants',
    'writing-poetry': 'Poetry',
    'writing-proverbs': 'Proverbs',
    'writing-quotations': 'Quotations and Quote Margins',
  },
  numbers: {
    'translate-fraction': 'Fractions',
    'translate-numbers': 'Numbers',
    'translate-ordinal': 'Ordinal Numbers',
  },
  figures: {
    'figs-apostrophe': 'Apostrophe',
    'figs-doublenegatives': 'Double Negatives',
    'figs-doublet': 'Doublet',
    'figs-ellipsis': 'Ellipsis',
    'figs-euphemism': 'Euphemism',
    'figs-hendiadys': 'Hendiadys',
    'figs-hyperbole': 'Hyperbole and Generalization',
    'figs-idiom': 'Idiom',
    'figs-irony': 'Irony',
    'figs-litotes': 'Litotes',
    'figs-merism': 'Merism',
    'figs-metaphor': 'Metaphor',
    'figs-metonymy': 'Metonymy',
    'figs-parallelism': 'Parallelism',
    'figs-personification': 'Personification',
    'figs-rquestion': 'Rhetorical Question',
    'figs-simile': 'Simile',
    'figs-synecdoche': 'Synecdoche',
    'figs-quotemarks': 'Quote Markings',
  },
  culture: {
    'figs-explicit': 'Assumed Knowledge and Implicit Information',
    'figs-go': 'Go and Come',
    'translate-bdistance': 'Biblical Distance',
    'translate-bmoney': 'Biblical Money',
    'translate-bvolume': 'Biblical Volume',
    'translate-hebrewmonths': 'Hebrew Months',
    'translate-names': 'How to Translate Names',
    'translate-bweight': 'Biblical Weight',
    'translate-symaction': 'Symbolic Action',
    'translate-unknown': 'Translate Unknowns',
    'writing-symlanguage': 'Symbolic Language',
  },
  grammar: {
    'figs-abstractnouns': 'Abstract Nouns',
    'figs-activepassive': 'Active or Passive',
    'figs-distinguish': 'Distinguishing versus Informing or Reminding',
    'figs-exclusive': 'Exclusive and Inclusive "We"',
    'figs-123person': 'First, Second, or Third Person',
    'figs-they': 'Forms of They',
    'figs-you': 'Forms of You',
    'figs-we': 'Forms of We',
    'figs-genericnoun': 'Generic Noun Phrases',
    'figs-hypo': 'Hypothetical Situations',
    'figs-inclusive': 'Inclusive "We"',
    'figs-nominaladj': 'Nominal Adjectives',
    'figs-possession': 'Possession',
    'figs-pronouns': 'Pronouns',
    'figs-rpronouns': 'Reflexive Pronouns',
    'figs-gendernotations': 'When Masculine Words Include Women',
    'grammar-connect-logic-goal': 'Connect - Goal (Purpose) Relationship',
  },
  other: {
    'guidelines-sonofgodprinciples': 'Translating Son and Father',
    'translate-manuscripts': 'Original Manuscripts',
    'translate-textvariants': 'Textual Variants',
    'translate-transliterate': 'Copy or Borrow Words',
  },
};

/**
 * Gets the category name of a groupId. Returns other as a categpory name
 * if the groupId is not found in the tNotes categories in the list above.
 * @param {string} groupId tA article slug.
 * @returns {string} The category name of a groupId: discourse,
 * numbers, figures, culture, grammar or other.
 */
export const getCategoryForGroupId = groupId => {
  let result = 'other';

  Object.keys(tNotesCategories).forEach(category => {
    if (tNotesCategories[category][groupId]) {
      result = category;
    }
  });

  return result;
};

/**
 * Categorizes the group data into five categories: discourse,
 * numbers, figures, culture, grammar or other.
 * @param {Object} groupData Object of groupids that contain an
 * array of contextIds for each groupId.
 * @returns The group data categorized into five categories:
 * discourse, numbers, figures, culture, grammar or other.
 */
export const categorizeGroupData = groupData => {
  const categorizedGroupData = {
    discourse: {},
    numbers: {},
    figures: {},
    culture: {},
    grammar: {},
    other: {},
  };

  Object.keys(groupData).forEach(groupId => {
    const category = getCategoryForGroupId(groupId);
    categorizedGroupData[category][groupId] = groupData[groupId];
  });

  return categorizedGroupData;
};
