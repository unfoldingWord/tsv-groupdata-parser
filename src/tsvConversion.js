
/**
 * conver array of Reference chunks to reference string
 * @param {string} tsv - data in 9 column format
 * @return {string}
 */
export function convertTsv9to7(tsv) {
  let tsvObjects;

  try {
    tsvObjects = tsvtojson_(tsv);
  } catch (e) {
    console.error(`convertTsv9to7() - could not convert TSV data to json"`, e);
    tsvObjects = null;
  }

  if (Array.isArray(tsvObjects)) {
    ??
  } else {
    console.error(`convertTsv9to7() - invalid tsv data"`);
    tsvObjects = null;
  }

  return tsvObjects;
}

/**
 * convert tsv data to json objects
 *  taken from tsvtojson - with fs.readFileSync removed
 * @param tsv
 * @param headers
 * @param splitString
 * @returns {*[]}
 */
export const tsvtojson_ = (tsv, headers, splitString) => {
  let header = headers || [];
  let json = [];
  let splitString_ = splitString || '\n';
  tsv.split(splitString_).forEach((line, index) => {
    if (!index && !header.length) {
      header = line.split('\t');
    } else {
      let obj = {};
      line.split('\t').forEach((value, index) => {
        value ? (obj[header[index]] = value) : '';
      });
      Object.keys(obj).length ? json.push(obj) : '';
    }
  });
  return json;
};
