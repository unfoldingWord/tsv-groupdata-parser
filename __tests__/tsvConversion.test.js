import fs from 'fs-extra';
import path from 'path-extra';
// helpers
import { convertTsv9to7 } from '../src/tsvConversion';

jest.unmock('fs-extra');

// constants
const TSV_PATH = path.join(__dirname, 'fixtures/tsv');

describe('Tests convertTsv9to7', function () {
  const fileNames = ['en_tn_08-RUT.tsv', 'en_tn_41-MAT.tsv', 'en_tn_42-MRK.tsv', 'en_tn_57-TIT.tsv', 'hi_tn_57-TIT.tsv'];

  for (const fileName of fileNames) {
    it(`${fileName} should pass`, () => {
      // given
      const tsvPath = path.join(TSV_PATH, fileName);
      const tsv = fs.readFileSync(tsvPath, 'utf8');
      const expectedLines = getLineCount(tsv);

      // when
      const results = convertTsv9to7(tsv);

      //then
      expect(results.tsv).toBeTruthy();
      const resultsLines = getLineCount(results.tsv);
      expect(resultsLines).toEqual(expectedLines);
      expect(results.errors).toMatchSnapshot();
    });
  }
});

// Helpers

function getLineCount(tsv) {
  const lines = tsv.split('\n');
  let numLines = lines.length;

  if (!lines[numLines - 1]) {
    numLines--;
  }
  return numLines;
}

