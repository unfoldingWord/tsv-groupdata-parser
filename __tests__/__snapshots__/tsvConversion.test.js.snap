// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests convertTsv9to7 ID tests should catch invalid chars in ID 1`] = `"TIT 1:1 row ID has invalid 1 character R, in line 2: TIT	1	1	xRpz	figs-abstractnouns	ἐπίγνωσιν ἀληθείας	1	knowledge of the truth	**knowledge** and **truth** are abstract nouns. See the UST for other ways to express these. Paul wants people to know the true message about God and Christ so that they can live in a way that pleases God. (See: [[rc://en/ta/man/translate/figs-abstractnouns]])"`;

exports[`Tests convertTsv9to7 ID tests should catch invalid first char in ID 1`] = `"TIT 1:1 row ID has invalid first character 0, in line 2: TIT	1	1	0xyz	figs-abstractnouns	ἐπίγνωσιν ἀληθείας	1	knowledge of the truth	**knowledge** and **truth** are abstract nouns. See the UST for other ways to express these. Paul wants people to know the true message about God and Christ so that they can live in a way that pleases God. (See: [[rc://en/ta/man/translate/figs-abstractnouns]])"`;

exports[`Tests convertTsv9to7 ID tests should catch short ID 1`] = `"Expected TIT 1:1 row ID to be 4 characters (not 3 characters with 'xyz'), in line 2: TIT	1	1	xyz	figs-abstractnouns	ἐπίγνωσιν ἀληθείας	1	knowledge of the truth	**knowledge** and **truth** are abstract nouns. See the UST for other ways to express these. Paul wants people to know the true message about God and Christ so that they can live in a way that pleases God. (See: [[rc://en/ta/man/translate/figs-abstractnouns]])"`;

exports[`Tests convertTsv9to7 OrigQuote tests should correct spaces in OrigQuote 1`] = `""`;

exports[`Tests convertTsv9to7 OrigQuote tests should replace ... with & 1`] = `""`;

exports[`Tests convertTsv9to7 OrigQuote tests should replace ellipsis with & 1`] = `""`;

exports[`Tests convertTsv9to7 Test OrigQuote replace html break 1`] = `""`;

exports[`Tests convertTsv9to7 empty support ref 1`] = `"Expected no OrigQuote for occurrence=='0' for TIT 1:1  'ἐπίγνωσιν ἀληθείας' 0 'knowledge of the truth', in line 2: TIT	1	1	wxyz		ἐπίγνωσιν ἀληθείας	0	knowledge of the truth	**knowledge** and **truth** are abstract nouns. See the UST for other ways to express these. Paul wants people to know the true message about God and Christ so that they can live in a way that pleases God. (See: [[rc://en/ta/man/translate/figs-abstractnouns]])"`;

exports[`Tests convertTsv9to7 en_tn_08-RUT.tsv should pass 1`] = `""`;

exports[`Tests convertTsv9to7 en_tn_41-MAT.tsv should pass 1`] = `"Expected occurrence=='0' for MAT 13:13 rc://*/ta/man/translate/figs-parallelism '' 1 'Though they are seeing, they do not see; and though they are hearing, they do not hear, or understand.', in line 1318: MAT	13	13	uk7j	figs-parallelism		1	Though they are seeing, they do not see; and though they are hearing, they do not hear, or understand.	Jesus uses this parallelism to tell and emphasize to the disciples that the crowd refuses to understand God’s truth. (See: [[rc://en/ta/man/translate/figs-parallelism]])"`;

exports[`Tests convertTsv9to7 en_tn_42-MRK.tsv should pass 1`] = `
"Expected no OrigQuote for occurrence=='0' for MRK 2:19 rc://*/ta/man/translate/figs-rquestion 'μὴ δύνανται οἱ υἱοὶ τοῦ νυμφῶνος ἐν ᾧ ὁ νυμφίος μετ’ αὐτῶν ἐστιν νηστεύειν?' 0 'The wedding attendants cannot fast while the bridegroom is still with them, can they?', in line 130: MRK	2	19	eke3	figs-rquestion	μὴ δύνανται οἱ υἱοὶ τοῦ νυμφῶνος ἐν ᾧ ὁ νυμφίος μετ’ αὐτῶν ἐστιν νηστεύειν?	0	The wedding attendants cannot fast while the bridegroom is still with them, can they?	Jesus uses this question to remind the people of something they already know and to encourage them to apply it to him and his disciples. Alternate translation: “Wedding attendants do not fast while the bridegroom is with them. Rather they celebrate and feast.” (See: [[rc://en/ta/man/translate/figs-rquestion]])
Expected no OrigQuote for occurrence=='0' for MRK 8:14 rc://*/ta/man/translate/figs-litotes 'εἰ μὴ ἕνα ἄρτον' 0 'except for one loaf', in line 603: MRK	8	14	gtg6	figs-litotes	εἰ μὴ ἕνα ἄρτον	0	except for one loaf	The negative phrase “no more” is used to emphasize how small an amount of bread they had. Alternate translation: “only one loaf” (See: [[rc://en/ta/man/translate/figs-litotes]])
Expected no OrigQuote for occurrence=='0' for MRK 8:33 rc://*/ta/man/translate/figs-metaphor 'ὕπαγε ὀπίσω μου, Σατανᾶ, ὅτι οὐ φρονεῖς' 0 'Get behind me, Satan! For you are not setting your mind', in line 636: MRK	8	33	nu32	figs-metaphor	ὕπαγε ὀπίσω μου, Σατανᾶ, ὅτι οὐ φρονεῖς	0	Get behind me, Satan! For you are not setting your mind	Jesus means that Peter is acting like Satan because Peter is trying to prevent Jesus from accomplishing what God sent him to do. Alternate translation: “Get behind me, because you are acting like Satan! You are not setting” (See: [[rc://en/ta/man/translate/figs-metaphor]])"
`;

exports[`Tests convertTsv9to7 en_tn_57-TIT.tsv should pass 1`] = `""`;

exports[`Tests convertTsv9to7 en_tn_57-TIT-invalid-ref.tsv should pass 1`] = `
"NOTE: TIT 1:1 4 OccurrenceNote is empty: ', in line 4: TIT	1	1	rtc9		κατὰ πίστιν	1	for the faith
NOTE: TIT 1:3 10 OccurrenceNote is empty: ', in line 10: TIT	1	3	swi9	figs-metaphor	ἐφανέρωσεν…τὸν λόγον αὐτοῦ	1	he revealed his word"
`;

exports[`Tests convertTsv9to7 en_tn_57-TIT-missing-occurrence-note.tsv should pass 1`] = `
"NOTE: TIT 1:1 4 OccurrenceNote is empty: ', in line 4: TIT	1	1	rtc9		κατὰ πίστιν	1	for the faith
NOTE: TIT 1:1 5 OccurrenceNote is empty: ', in line 5: TIT	1	1	xyz8	figs-abstractnouns	ἐπίγνωσιν ἀληθείας	1	knowledge of the truth
NOTE: TIT 1:3 10 OccurrenceNote is empty: ', in line 10: TIT	1	3	swi9	figs-metaphor	ἐφανέρωσεν…τὸν λόγον αὐτοῦ	1	he revealed his word"
`;

exports[`Tests convertTsv9to7 en_tn_57-TIT-white-space-as-support-ref.tsv should pass 1`] = `
"NOTE: TIT 1:1 4 OccurrenceNote is empty: ', in line 4: TIT	1	1	rtc9		κατὰ πίστιν	1	for the faith
NOTE: TIT 1:1 5 OccurrenceNote is empty: ', in line 5: TIT	1	1	xyz8	figs-abstractnouns	ἐπίγνωσιν ἀληθείας	1	knowledge of the truth
NOTE: TIT 1:3 10 OccurrenceNote is empty: ', in line 10: TIT	1	3	swi9	figs-metaphor	ἐφανέρωσεν…τὸν λόγον αὐτοῦ	1	he revealed his word"
`;

exports[`Tests convertTsv9to7 hi_tn_57-TIT.tsv should pass 1`] = `""`;

exports[`Tests convertTsv9to7 replace double spaces 1`] = `
"NOTE: TIT 1:1 2 OccurrenceNote has unexpected double-spaces: '* Nuts.
**knowledge** and **truth** are abstract nouns. See the UST for other ways to express these.
  Paul wants people to know the true message about God and Christ so that they can live in a way that pleases God. (See: [[rc://*/ta/man/translate/figs-abstractnouns]]), in line 2: TIT	1	1	wxyz	figs-abstractnouns	ἐπίγνωσιν ἀληθείας	-1	knowledge of the truth	*  Nuts.<br>**knowledge** and **truth** are abstract nouns.  See the UST for other ways to express these.<br>  Paul wants people to know the true message about God and Christ so that they can live in a way that pleases God. (See: [[rc://en/ta/man/translate/figs-abstractnouns]])"
`;

exports[`Tests convertTsv9to7 should warn on not zero occurrence but don't have original quote 1`] = `"Expected occurrence=='0' for TIT 1:1 rc://*/ta/man/translate/figs-abstractnouns '' 1 'knowledge of the truth', in line 2: TIT	1	1	wxyz	figs-abstractnouns		1	knowledge of the truth	**knowledge** and **truth** are abstract nouns. See the UST for other ways to express these. Paul wants people to know the true message about God and Christ so that they can live in a way that pleases God. (See: [[rc://en/ta/man/translate/figs-abstractnouns]])"`;

exports[`Tests convertTsv9to7 should warn on zero occurrence but have original quote 1`] = `"Expected no OrigQuote for occurrence=='0' for TIT 1:1 rc://*/ta/man/translate/figs-abstractnouns 'ἐπίγνωσιν ἀληθείας' 0 'knowledge of the truth', in line 2: TIT	1	1	wxyz	figs-abstractnouns	ἐπίγνωσιν ἀληθείας	0	knowledge of the truth	**knowledge** and **truth** are abstract nouns. See the UST for other ways to express these. Paul wants people to know the true message about God and Christ so that they can live in a way that pleases God. (See: [[rc://en/ta/man/translate/figs-abstractnouns]])"`;

exports[`Tests convertTsv9to7 test invalid tsv's should fail should fail if no header 1`] = `"convertTsv9to7() - invalid tsv data, could not parse"`;

exports[`Tests convertTsv9to7 test invalid tsv's should fail should fail insufficient columns 1`] = `
"missing field 'OccurrenceNote' in header
Number of columns should be 9, but 8 columns found, in line 1: Book	Chapter	Verse	ID	SupportReference	OrigQuote	Occurrence	GLQuote"
`;

exports[`Tests convertTsv9to7 warn on empty OccurrenceNote 1`] = `"NOTE: TIT 1:1 2 OccurrenceNote is empty: ', in line 2: TIT	1	1	wxyz	figs-abstractnouns	ἐπίγνωσιν ἀληθείας	-1	knowledge of the truth	"`;
